### PR TITLE
Adding torch.nn.PReLU

### PR DESCRIPTION
**Pull request by skutaada**
Thank you for the repo! Already saved me a lot of time. I had a model that contained `torch.nn.PReLU`, so I wanted to add it.

### CODE REVIEW

File: torch2jax/__init__.py [hunk 1]
```diff
@@ -916,6 +916,24 @@ def relu(x, inplace=False):
...
+      f"Mismatch of parameter numbers and input channel size. Found parameter numbers = {weight.numel()} and channel size = {channel_size}."
+    )
+  assert weight.ndim == 0 or weight.ndim == 1, (
+    f"prelu: Expected `weight` to be a scalar or 1D tensor, but got: ndim = {weight.ndim}"
+  )
+  if a.ndim == 0:
+    weight.value = weight[0] if weight.ndim == 1 else weight.value
+  else:
+    weight.value = jax.lax.broadcast_in_dim(_v(weight), a.shape, () if weight.ndim == 0 else (0 if a.ndim == 1 else 1,))
+  return jnp.where(_v(a) > 0, _v(a), _v(a) * _v(weight))
```
Comment (samuela) [inline torch2jax/__init__.py:934]: why not just do

```suggestion
  return jnp.maximum(x, 0) + weight * jnp.minimum(x, 0)
```

does that break anything?

File: tests/test_torch_nn.py [hunk 2]
```diff
@@ -184,6 +184,28 @@ def test_torch_nn_Dropout():
   assert jnp.mean(res_jax == res_torch.numpy(force=True)) > 0.5
 
 
+def test_torch_nn_PReLU():
```
Comment (samuela) [inline tests/test_torch_nn.py:187]: nit: let's keep these tests alphabetized, so this would go after `test_torch_nn_MaxPool2d`

File: tests/test_torch_nn.py [hunk 3]
```diff
@@ -184,6 +184,28 @@ def test_torch_nn_Dropout():
...
+
+  # Test forward pass without and with jit
+  aac(res_jax, res_torch.numpy(force=True), atol=1e-5)
+  aac(res_jax_jit, res_torch.numpy(force=True), atol=1e-5)
+
+  # Test gradients
+  jax_grad = grad(lambda p: (jaxified_module(input_batch, state_dict=p) ** 2).sum())(params)
+
+  res_torch.pow(2).sum().backward()
+  aac(jax_grad["weight"], model.weight.grad, atol=1e-3)
```
Comment (samuela) [inline tests/test_torch_nn.py:206]: let's also test the gradient on `input`. is there a clean way to test that here? if not you can test `torch.nn.functional.prelu` directly and `t2j_function_test` will handle it for you

File: torch2jax/__init__.py [hunk 4]
```diff
@@ -916,6 +916,24 @@ def relu(x, inplace=False):
     return Torchish(jax.nn.relu(_v(x)))
 
 
+@implements(torch.nn.functional.prelu)
+def prelu(a: Torchish, weight: Torchish):
```
Comment (samuela) [inline torch2jax/__init__.py:920]: pytorch calls the first argument "input": https://docs.pytorch.org/docs/stable/generated/torch.nn.functional.prelu.html

Would this work with eg `t2j(lambda x: torch.nn.functional.prelu(input=x, weight=123))`?

**Review by samuela — CHANGES_REQUESTED** [review]
Thanks @skutaada !

File: torch2jax/__init__.py (hunk 1 above)
Comment (skutaada) [reply torch2jax/__init__.py:934]: The broadcasting would not work. For example the test I wrote would fail. This is basically a direct rewrite from PyTorch implementation of prelu (https://github.com/pytorch/pytorch/blob/main/torch/_refs/nn/functional/__init__.py#L1125).

File: torch2jax/__init__.py [hunk 5]
```diff
@@ -916,6 +916,26 @@ def relu(x, inplace=False):
...
+  assert weight.ndim == 0 or weight.ndim == 1, (
+    f"prelu: Expected `weight` to be a scalar or 1D tensor, but got: ndim = {weight.ndim}"
+  )
+  if input.ndim == 0:
+    weight.value = weight[0] if weight.ndim == 1 else weight.value
+  else:
+    weight.value = jax.lax.broadcast_in_dim(
+      _v(weight), input.shape, () if weight.ndim == 0 else (0 if input.ndim == 1 else 1,)
+    )
+  return jnp.where(_v(input) > 0, _v(input), _v(input) * _v(weight))
```
Comment (samuela) [inline torch2jax/__init__.py:936]:
```suggestion
  if input.ndim == 0:
    weight = weight[0] if weight.ndim == 1 else weight
  else:
    weight = jax.lax.broadcast_in_dim(
      _v(weight), input.shape, () if weight.ndim == 0 else (0 if input.ndim == 1 else 1,)
    )
  return jnp.where(_v(input) > 0, _v(input), _v(input) * _v(weight))
```
I'm concerned that mutating `weight` could have some unintended consequences, although I can't put my finger on a test case atm. In general it's preferable to avoid messing with `Torchish` internals unless strictly necessary.

**Review by samuela — CHANGES_REQUESTED** [review]
One small thing and then I think this is ready to go

File: torch2jax/__init__.py (hunk 1 above)
Comment (samuela) [reply torch2jax/__init__.py]: Thanks, wasn't aware of that

File: torch2jax/__init__.py (hunk 5 above)
Comment (skutaada) [reply torch2jax/__init__.py]: Changed it. Needed to wrap in Torchish on line 939 for the return to make sense.
